### PR TITLE
feat: add sign() method

### PR DESCRIPTION
**Pull request by iamssen**
# TODO

- [x] Add `sign()` method
- [x] Implement Chrome Extension
- [ ] Implement Terra Station Mobile
- [ ] Implement Next Web Extension
- [ ] Update CRA template
- [ ] Update Next template
- [ ] Update Controller template

Related <https://github.com/terra-money/wallet-provider/issues/4>

# Examples

- <https://github.com/terra-money/wallet-provider/blob/feature%2Fsign/packages/src/components/SignSample.tsx>

# Install test version

```sh
yarn add @terra-money/wallet-provider@next # or @terra-money/wallet-provider@1.3.0-alpha.1
```

See <https://www.npmjs.com/package/@terra-money/wallet-provider/v/next>

### CODE REVIEW

**Comment by cryptoethic** [comment]
Thanks for start working on that feature. I played a little bit with SignSample you provided and I have two questions related to this part:

```tsx
    connectedWallet
      .sign({
        fee: new StdFee(1000000, '200000uusd'),
        msgs: [
          new MsgSend(connectedWallet.walletAddress, toAddress, {
            uusd: 1000000,
          }),
        ],
      })
```


1. Is there a message type that allow you pass any string data? If not is it worth adding? As far as I understand message type is further used to display it correctly in the UI? Closest message I have found is MsgStoreCode - it allows me to send any string.
2. In our case we would like to sign message without broadcasting it to the network, but send it directly to our backend and then verify signature. Therefore fee parameter from CreateTxOptions shouldn't be needed. In current example removing fee property causes `TxUnspecifiedError` with no message.

**Comment by iamssen** [comment]
@cryptoethic I don't know much about Terra Blockchain system because I'm developing Frontend.

I don't think I can give you any technical guide on Contract.

However, I think you will always need a gas feed to provide to Chain in order to execute Tx. I understand that there is no way to use Chain without providing Fee because it is a Fee for processing Tx.

**Comment by cryptoethic** [comment]
Thanks for fast reply. Gas fee is not a big problem at this moment. 

So if i could help with implementing sing method for Next Web Extension and Terra Station Mobile - where could i find this Next Web Extension? 

I could not locate source code for npm package @terra-dev/web-extension. WebExtensionController does not provide sign message so I believe this package needs to be updated. 

In case of walletconnect package I can see signMessage and signPersonalMessage methods available so it might be easier to implement

**Comment by kjessec** [comment]
@cryptoethic May I ask what you're trying to do with your question 1? Message types are predefined types in our network and it must be one of the supported msgs. If you're just trying to leave a remark in your transaction you could use the `memo` option.

**Comment by cryptoethic** [comment]
Hi @kjessec!

So basically we are trying achieve sign custom message functionality. That way we could ask a wallet owner to sign a message for us (for example session id) and give it back to our backend server. Later we can validate signature and confirm that this wallet address truly belongs to user with specified session id. 

We are working on such solution to ensure that users doing KYC process are owners of their wallets.

---

In web3 library there is this method: [web3.eth.sign](https://web3js.readthedocs.io/en/v1.2.9/web3-eth.html#sign)

---

So basically we are not talking about signing transaction there but any data without broadcasting it to the network. If it comes to the memo option - I believe it is limited to [256 bytes](https://docs.terra.money/dev/spec-auth.html#maxmemocharacters)? 

Creating a signed transaction, and instead of broadcasting to the network transferring to backend server is a kind of workaround but could do a job for our use case. Unfortunately there is no predefined msg type for this use case.

**Comment by kjessec** [comment]
@cryptoethic so you're letting a user to sign arbitrary byte with the private key? That is certainly not the scope of this project I'm afraid.

As for the web3 sign thing you linked -- it is more of a standard byte-signer function. And while we do have similar happening underneath, it still is not the scope of this project (or the terra.js library for that matter).

**Comment by cryptoethic** [comment]
Thanks for your response @kjessec.

As you said you have tried KYC verification yourself. Maybe there are some conclusions worth sharing?

If signing arbitrary data is not scope of terra.js nor wallet-provider than I believe I should focus on signing transaction with message playload that contains my data.

Use case would look like this:
> User creates signed transaction with custom message, where payload is arbitrary data (that can be done currently with Chrome Extension, and needs to be implemented for Terra Mobile and Next Web Extension)
> Instead of broadcasting that transaction to the network we are sending it to our backend server
> Backend server validates signatures and extract signed data

This allows us to sign any data, but as you said before all transactions must consist of predefined types of messages. Hence my question of adding another message type that would be open for future use cases like MsgExtra that accept any data (or string).

I would love to hear if you come up with any different approaches for KYC address ownership verification.

**Comment by kjessec** [comment]
@cryptoethic

1. we're not planning to add any MsgExtra-ish message types -- it kind of requires a hardfork on our end and our blockchain spec is finalised as of now 😅.
2. I still believe the best way to do this is to let the user directly sign an arbitrary payload and send it directly to your backend, but unfortunately for now we're not going to provide such hook features on our wallet projects (despite there is a discussion going on).
3. Having said all this, the only workaround would be to use the blockchain as message relayer -- either using the aformentioned memo field, or setup a contract which accepts a token that your server would have issued, then verify the signature within the contract -- and your backend can monitor successful transactions off-chain. Detailed verification process is to be explored separately, but given our airdrop contracts do merkle tree verification, I believe any verification method would be possible.

**Comment by kjessec** [comment]
Despite all I have said we might actually add this functionality in the future. I will keep you posted on this thread.

**Comment by kjessec** [comment]
@cryptoethic I guess we're adding it

**Comment by cryptoethic** [comment]
@kjessec  That is a great news ;-)
When should we expect this functionality? On Col 5 start or later?
